README
////
execute as "asciidoctor manual.txt"
////

butt (0.1.30) Manual 
====================
:author: Daniel Nöthen
:doctype: book 
:toc2:
:numbered:
:lang: en
:email: butt at danielnoethen dot de
:encoding: utf-8

About
-----
butt (broadcast using this tool) is an easy to use, multi OS streaming tool. +
It supports ShoutCast and IceCast and runs on Linux, MacOS and Windows. +
The main purpose of butt is to stream live audio data from your computers Mic or Line input +
to an Shoutcast or Icecast server. Recording is also possible. +
It is NOT intended to be a server by itself or automatically stream a set of audio files. +


Install
-------
.OS X: 
Mount the *butt-<version>.dmg* by double clicking and  +
drop the butt file to the Applications folder. +

.Windows: 
Just run the *butt-<version>-setup.exe* as usual and go through the +
installer pages. +
The default installation path might be unusual, but this way it is possible to install +
butt without administration rights. +

.Linux/MinGW (Windows):
First of all the following libraries have to be installed on your system  +
'fltk-1.3', 'portaudio19', 'libmp3lame', 'libvorbis', 'libogg', 'libflac', 'libopus', 'libsamplerate', 'libfdk-aac', 'libdbus-1, libssl' +
They are quite common and should be included in every popular linux distribution. +

On *Ubuntu* you can install them with +
`sudo apt-get install libfltk1.3-dev portaudio19-dev libopus-dev libmp3lame-dev libvorbis-dev 
libogg-dev libflac-dev libfdk-aac-dev libdbus-1-dev libsamplerate0-dev libssl-dev`

On Distributions which don't have libfdk-aac like *debian* 
you can compile without aac support with: +
`./configure --disable-aac` 

If you want to disable SSL/TLS support you can run: +
`./configure --disable-ssl` 


For compiling on *Windows* I recommend the msys2 x64 (www.msys2.org) environment. +
They have all the needed packages in their pacman repository. Additionally you need to install the `libwinpthread-git` package. +
Make sure that you select the x64 version of each package.

After installing the above libraries you can install butt from source as usual:

- +tar -xzf butt-<version>.tar.gz+

- +cd butt-<version>+

- +autoreconf -i+ (only on Windows/MSYS2)

- +./configure+

- +make+

- +sudo make install+

In case the included configure script or the make process fails on your system, try to create a new configure script by invoking:
`autoreconf -i`
and start with  `./configure` again.



Quick start
-----------
When you start butt the first time, it will create a default configuration +
file in in your home directory ('~/.buttrc') on Linux and OS X or at +
'C:\Users\<username>\AppData\Roaming\buttrc' on Windows. +

In order to connect to a server, you need to add a new server in the config window.  +
Just open the settings window and click on [ADD]. +
Now fill in the input fields with the server data and click on the new [ADD]. +
To connect to the server just press the play button in the main window and enjoy +
your broadcast.

Install AAC libraries
---------------------
.Windows:
1. Download libfdk-aac-2.dll from
https://sourceforge.net/projects/butt/files/butt/butt-0.1.30/AAC/libfdk-aac-2.dll/download[here]
2. Go to the installation directory of butt by typing %LOCALAPPDATA%\butt into the file explorer 
3. Copy libfdk-aac-2.dll into the installation directory of butt
4. Restart butt
5. Enjoy AAC Streaming

.MacOS: 
1. Download libfdk-aac.2.dylib from 
https://sourceforge.net/projects/butt/files/butt/butt-0.1.30/AAC/libfdk-aac.2.dylib/download[here] +
2. Open your Applications folder +
3. Right click on butt.app +
4. Select Show Package Contents +
5. Copy libfdk-aac.2.dylib next to the butt executeable in Contents/MacOS +
6. Restart butt
7. Enjoy AAC Streaming +

.Linux:
1. Install libfdk-aac from your linux distribution repository
2. Enjoy AAC Streaming



Configuration
-------------
The command line option +-c <path_to_file>+ allows you to define a new standard configuration path.  +
This makes it possible to have multiple instances with different configurations +
running. In case the file does not exists, butt will create a default file. +

[Save]: Saves your current settings to the standard configuration +
file or to the file that was passed to the -c option +
[Export]: Saves your current settings to the given file +
[Import]: Loads the selected file and applies the settings +

CAUTION: If you use the -c command line option and import another configuration file + 
by using the import function, pressing [Save] will overwrite the file that was passed to the -c option.

In some cases the configuration file can become corrupted and butt will not be
able to start correctly. In that case you can hold down the shift key during
butt startup to create a new configuration file. 


Main Window
-----------
The dot matrix display shows you the current state of the butt software. +
The states are: idle, streaming, recording. +
When in streaming and/or recording state you can cycle through the information by clicking on the display. +
You can choose between online duration, data sent, recording duration and data recorded +

The [>] symbol shines yellow if butt is connected to a server. +
The [O] symbol shines orange if the +[start rec. when connected]+ checkbox is activated. +
The [O] symbol shines red if butt is currently recording. +

Gain slider:
The slider is only visible when the little +[more/less]+ button below  +
the +[settings]+ button was clicked. +
With this slider you can attenuate and amplify the input signal between '-24 dB' and  '+24 dB', respectively. +
Double clicking the slider resets the gain to '0 dB'. Use this slider only to fine tune your input signal.  +
It does not change the operating systems input volume setting. Instead, the input signal is multiplied +
by the given factor. Thus, adding too much gain will also add lots of noise. +

Audio
-----
The audio settings tab allows you to select your audio interface and the desired sample rate. +
If you have a multi-channel device you may also select the desired input channels for the left and + 
right audio channel. +

With the channel mode setting you tell butt if the audio stream shall be encoded to stereo or mono. +
This brings us to 4 possible combinations.

.Channel mode = *Stereo* and *different channels* are selected for Left and Right: 
This is probably the most used combination. +
Left and Right channels are encoded into a stereo stream.

.Channel mode = *Stereo* and the *same channel* is selected for Left and Right:
The selected channel is used for Left and Right channel in a stereo stream.

.Channel mode = *Mono* and *different channels* are selected for Left and Right:
Left and Right channels are averaged into a mono stream. 

.Channel mode = *Mono* and the *same channel* is selected for Left and Right:
The selected channel is used as mono channel in a mono stream.


Streaming
---------
To start streaming just click the play symbol. +
butt will try to connect to the server until you press the stop button. +
If the connection gets lost, butt will try to reconnect until the stop button is pressed +

You can stream in 5 different audio codecs: mp3, aac+, ogg/vorbis, ogg/opus and FLAC. + 
In case opus is selected the sample rate is always upsampled to 48 kHz. +
Of course no upsampling is needed if you select 48 kHz as sample rate. +
  
.Song name:
If you want to inform the listener about which song is currently playing you +
can do that on the +[Stream]+ tab. +
You only need to type the song into the +Song Name+ input field +
and hit Enter or click +[OK]+. 

butt can also update the song automatically from a text file. +
The first or the last line of the file must be the name of the song. +
As soon as butt detects that the file has been changed, it updates the +
name of the song on the server. +

If you run butt on MacOS or Linux you can even transfer the current +
song name from an audio player to butt. +
Supported audio players: +
Linux: Rhythmbox, Banshee, Clementine, Cantana, Spotify +
MacOS: iTunes/Music, Vox, Spotify

In case you want to add a prefix and/or a suffix to your song name +
you can do that by entering the desired text into the corresponding +
input field.

Updating the song name is not supported for the Opus codec.
 
.Stream infos:
In the +[Main]+ settings window you can add stream infos. +
This allows you to deliver more details about your stream. +
For example the genre of your music, description of your station, web address etc. +

Unfortunately, it is not possible to update stream infos during a broadcast. +
You need to reconnect for updating the stream infos. +

.Automatic streaming:
If you activate the checkbox 'Start streaming after launch' butt will +
automatically connect to the server as soon as the application has been started.

butt can also connect and disconnect depending on the audio signal level: +

To connect automatically if a signal is present for a certain amount of time +

ChangeLog

Version 0.1.30 (2021-05-14)
---------------------------
* Added support for IPv6 
* Added French translation
* Right click on a slider knob brings up an input field where you can type in the slider value with the keyboard
* The selected audio device can now be remembered by device name instead by device id
* You can now rescan for new audio devices manually (no butt restart needed anymore after plugging in a new audio device)
* Holding the shift key during butt startup allows you to start with a fresh configuration file
* Updated Portaudio to version v19.7.0
* Windows: Added support for WINAPI audio driver
* MacOS: Updated FLTK to version 1.3.6 
* MacOS: Native support for Apple Silicon (arm64) 

Version 0.1.29 (2021-03-28)
---------------------------
* Added a new tool (butt Agent) which helps to minimize butt to tray and much more (Big thanks to Ivan Radolovic)
* Added an indicator to the compressor which lights up if the threshold is exceeded
* Added "Aggressive Mode" to the compressor. If activated, the signal power for threshold detection is not averraged
* Added new command line options: -q (quit) -n (split recording) -u <songname> (update song name)
* Added donation addresses for Bitcoin, Monero and Dash
* Fixed crash when butt can not find any input channels
* Fixed crash when reading song names from an UTF-8 formatted file that contains a BOM 
* Fixed issue with icecast-kh14+ when connecting to a mountpoint which is already in use
* Fixed issue if both checkboxes "start recording after launch" and "start recording when connected" are checked
* Fixed butt.desktop file. Icon was not shown on GNOME panel
* Attack and Release compressor labels now have 2 decimal places instead of only 1
* butt does not send bitrate information to the icecast server anymore if FLAC is used 
* butt now searches for a free command server port if default port is already in use
* butt can now be started in minimized state
* The recording file name can now contain any well known date format specifier 
* The window class name has been renamed from FLTK to butt_FLTK

Version 0.1.28 (2021-02-07)
---------------------------
* Fixed compressor issues
* Fixed song update when using Strawberry audio player
* The order of song update from an music app can be changed from "Title - Artist" to "Artist - Title"
* Increased accuracy of recording split timer
* VU-meter appearance can be changed to solid colors for better readability
* Log window can be vertically resized
* Windows: Changed default installation folder to %LOCALAPPDATA%\butt instead of %LOCALAPPDATA\butt-<version>


Version 0.1.27 (2020-12-26)
---------------------------
* Fixed loading of AAC codec selection from configuration file
* Fixed crash when using opus with a samplerate not equal to 48000 Hz

Version 0.1.26 (2020-12-24)
---------------------------
* Added ASIO support
* Added support for multi-channel audio devices
* Added multi-language support (i18n) 
* Added German translation
* Added option to force reconnection regardless of the server response
* Added support for Strawberry audio player (Thanks to Stefan S.)
* Added config parameter to disable dithering (set disable_dithering=1 in cfg file) 
* Fixed possible crash if server is not reachable
* Fixed possible crash when reconnecting
* Fixed possible crash when using FLAC and song update
* Fixed unwanted rounding of compressor values
* Fixed overwriting of a DSP setting if it was set to -1.0
* Improved CPU usage 
* Improved mono/stereo handling
* Improved overall song update implementation
* Compiles on FreeBSD (Thanks to Per G. and Tobias C. B.)


Version 0.1.25 (2020-10-26)
---------------------------
* Added the ability to extend the song name with a prefix and a suffix
* Fixed issue with automatic song update from text files 
* Fixed recording in FLAC 

Version 0.1.24 (2020-10-14)
---------------------------
* VU-meter has been redesigned
* Fixed issues with usernames that contain special characters 
* The port of the command server can be specified via command line 
* Commands can be received via network
* It is possible to connect to a particular streaming server when sending the connect command
* You can ask a butt instance about its current status with the -S command
* butt can automatically check for updates after launch 
* Update of song name works with FLAC 
* Thresholds for signal detection can be specified
* Added -h command line option to get an overview of all available commands
* Updated the manual 

Version 0.1.23 (2020-09-27)

src/FLTK/fl_funcs.cpp
    // FLTK GUI related functions
//
// Copyright 2007-2018 by Daniel Noethen.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <signal.h>
#include <time.h>
#include <sys/stat.h>

#ifndef WIN32
 #include <sys/wait.h>
#endif

#include "gettext.h"
#include "config.h"

#include "cfg.h"
#include "butt.h"
#include "util.h"
#include "port_audio.h"
#include "timer.h"
#include "flgui.h"
#include "fl_funcs.h"
#include "shoutcast.h"
#include "icecast.h"
#include "strfuncs.h"


void fill_cfg_widgets(void)
{
    int i;

    int bitrate[] = { 8, 16, 24, 32, 40, 48, 56, 64, 80, 96, 112,
                    128, 160, 192, 224, 256, 320, 0 };

#ifndef HAVE_LIBFDK_AAC
    fl_g->menu_item_cfg_aac->hide();
    fl_g->menu_item_rec_aac->hide();
    Fl::check();
#endif
  
    //fill the main section
    for(i = 0; i < cfg.audio.dev_count; i++)
    {
        unsigned long dev_name_len = strlen(cfg.audio.pcm_list[i]->name)+10;
        char *dev_name = (char*)malloc(dev_name_len);
        
        snprintf(dev_name, dev_name_len, "%d: %s", i, cfg.audio.pcm_list[i]->name);
        fl_g->choice_cfg_dev->add(dev_name);
        free(dev_name);
    }

    
    fl_g->choice_cfg_dev->value(cfg.audio.dev_num);
    

    fl_g->choice_cfg_act_srv->clear();
    fl_g->choice_cfg_act_srv->redraw();
    for(i = 0; i < cfg.main.num_of_srv; i++)
        fl_g->choice_cfg_act_srv->add(cfg.srv[i]->name);

    if(cfg.main.num_of_srv > 0)
    {
        fl_g->button_cfg_edit_srv->activate();
        fl_g->button_cfg_del_srv->activate();
        fl_g->choice_cfg_act_srv->activate();
    }
    else
    {
        fl_g->button_cfg_edit_srv->deactivate();
        fl_g->button_cfg_del_srv->deactivate();
        fl_g->choice_cfg_act_srv->deactivate();
    }

    fl_g->choice_cfg_act_icy->clear();
    fl_g->choice_cfg_act_icy->redraw();
    for(i = 0; i < cfg.main.num_of_icy; i++)
        fl_g->choice_cfg_act_icy->add(cfg.icy[i]->name);

    if(cfg.main.num_of_icy > 0)
    {
        fl_g->button_cfg_edit_icy->activate();
        fl_g->button_cfg_del_icy->activate();
        fl_g->choice_cfg_act_icy->activate();

    }
    else
    {
        fl_g->button_cfg_edit_icy->deactivate();
        fl_g->button_cfg_del_icy->deactivate();
        fl_g->choice_cfg_act_icy->deactivate();
    }

    fl_g->choice_cfg_act_srv->value(cfg.selected_srv);
    fl_g->choice_cfg_act_icy->value(cfg.selected_icy);


    fl_g->check_cfg_connect->value(cfg.main.connect_at_startup);
    fl_g->check_cfg_force_reconnecting->value(cfg.main.force_reconnecting);

    fl_g->input_log_filename->value(cfg.main.log_file);

    fl_g->check_update_at_startup->value(cfg.main.check_for_update);
    
    fl_g->check_start_agent->value(cfg.main.start_agent);
    fl_g->check_minimize_to_tray->value(cfg.main.minimize_to_tray);


    //fill the audio section
    if(!strcmp(cfg.audio.codec, "mp3"))
        fl_g->choice_cfg_codec->value(CHOICE_MP3);
    else if(!strcmp(cfg.audio.codec, "ogg"))
        fl_g->choice_cfg_codec->value(CHOICE_OGG);
    else if(!strcmp(cfg.audio.codec, "opus"))
        fl_g->choice_cfg_codec->value(CHOICE_OPUS);
    else if(!strcmp(cfg.audio.codec, "aac"))
        fl_g->choice_cfg_codec->value(CHOICE_AAC);
    else if(!strcmp(cfg.audio.codec, "flac"))
    {
        fl_g->choice_cfg_codec->value(CHOICE_FLAC);
        fl_g->choice_cfg_bitrate->hide();

    }
    
    if (cfg.audio.dev_remember == REMEMBER_BY_ID)
        fl_g->radio_cfg_ID->setonly();
    else
        fl_g->radio_cfg_name->setonly();

    if(cfg.audio.channel == 1)
        fl_g->choice_cfg_channel->value(CHOICE_MONO);
    else
        fl_g->choice_cfg_channel->value(CHOICE_STEREO);

    for(i = 0; bitrate[i] != 0; i++)
    {
        if(cfg.audio.bitrate == bitrate[i])
        {
            fl_g->choice_cfg_bitrate->value(i);
            break;
        }
    }

    fl_g->input_cfg_present_level->value(-cfg.audio.signal_level);
    fl_g->input_cfg_absent_level->value(-cfg.audio.silence_level);

    fl_g->check_cfg_mono_to_stereo->value(cfg.audio.mono_to_stereo);
    
    fl_g->input_cfg_buffer->value(cfg.audio.buffer_ms);

    fl_g->choice_cfg_resample_mode->value(cfg.audio.resample_mode);
    
    // Fill stream section
    fl_g->check_song_update_active->value(cfg.main.song_update);
    fl_g->check_read_last_line->value(cfg.main.read_last_line);

    fl_g->input_cfg_song_file->value(cfg.main.song_path);
    fl_g->input_cfg_signal->value(cfg.main.signal_threshold);
    fl_g->input_cfg_silence->value(cfg.main.silence_threshold);
    
    fl_g->input_cfg_song_prefix->value(cfg.main.song_prefix);
    fl_g->input_cfg_song_suffix->value(cfg.main.song_suffix);

    
#if __APPLE__ && __MACH__
    fl_g->choice_cfg_app->add("iTunes\\/Music");
    fl_g->choice_cfg_app->add("Spotify");
    fl_g->choice_cfg_app->add("VOX");
    fl_g->check_cfg_use_app->value(cfg.main.app_update);
    fl_g->choice_cfg_app->value(cfg.main.app_update_service);
    if (cfg.main.app_artist_title_order == APP_ARTIST_FIRST)
        fl_g->radio_cfg_artist_title->setonly();
    else
        fl_g->radio_cfg_title_artist->setonly();
#elif (__linux__ || __FreeBSD__) && HAVE_DBUS
    fl_g->choice_cfg_app->add("Rhythmbox");
    fl_g->choice_cfg_app->add("Banshee");
    fl_g->choice_cfg_app->add("Clementine");
    fl_g->choice_cfg_app->add("Spotify");
    fl_g->choice_cfg_app->add("Cantata");
    fl_g->choice_cfg_app->add("Strawberry");
    fl_g->check_cfg_use_app->value(cfg.main.app_update);
    fl_g->choice_cfg_app->value(cfg.main.app_update_service);
    if (cfg.main.app_artist_title_order == APP_ARTIST_FIRST)
        fl_g->radio_cfg_artist_title->setonly();
    else
        fl_g->radio_cfg_title_artist->setonly();
#elif WIN32
    fl_g->choice_cfg_app->add(_("Not supported on Windows"));
    fl_g->choice_cfg_app->value(0);
    fl_g->check_cfg_use_app->value(0);
    fl_g->check_cfg_use_app->deactivate();
    fl_g->choice_cfg_app->deactivate();
    fl_g->radio_cfg_artist_title->deactivate();
    fl_g->radio_cfg_title_artist->deactivate();
#endif
    
    
    
    //fill the record section
    fl_g->input_rec_filename->value(cfg.rec.filename);
    fl_g->input_rec_folder->value(cfg.rec.folder);
    fl_g->input_rec_split_time->value(cfg.rec.split_time);

    if(!strcmp(cfg.rec.codec, "mp3"))
        fl_g->choice_rec_codec->value(CHOICE_MP3);
    else if(!strcmp(cfg.rec.codec, "ogg"))
        fl_g->choice_rec_codec->value(CHOICE_OGG);
    else if(!strcmp(cfg.rec.codec, "opus"))
        fl_g->choice_rec_codec->value(CHOICE_OPUS);
    else if(!strcmp(cfg.rec.codec, "aac"))
        fl_g->choice_rec_codec->value(CHOICE_AAC);
    else if(!strcmp(cfg.rec.codec, "flac"))
    {
        fl_g->choice_rec_codec->value(CHOICE_FLAC);
        fl_g->choice_rec_bitrate->hide();
    }
    else //wav
    {
        fl_g->choice_rec_codec->value(CHOICE_WAV);
        fl_g->choice_rec_bitrate->hide();
    }

    for(i = 0; bitrate[i] != 0; i++)
        if(cfg.rec.bitrate == bitrate[i])
            fl_g->choice_rec_bitrate->value(i);

    if(cfg.rec.start_rec)
        fl_g->check_cfg_auto_start_rec->value(1);
    else
        fl_g->check_cfg_auto_start_rec->value(0);
    
    if(cfg.rec.stop_rec)
        fl_g->check_cfg_auto_stop_rec->value(1);
    else
        fl_g->check_cfg_auto_stop_rec->value(0);
    

    if(cfg.rec.rec_after_launch)
        fl_g->check_cfg_rec_after_launch->value(1);
    else
        fl_g->check_cfg_rec_after_launch->value(0);

    if(cfg.rec.sync_to_hour)
        fl_g->check_sync_to_full_hour->value(1);
    else
        fl_g->check_sync_to_full_hour->value(0);

    fl_g->input_rec_signal->value(cfg.rec.signal_threshold);
    fl_g->input_rec_silence->value(cfg.rec.silence_threshold);

    //fill the ssl/tls section
    fl_g->input_tls_cert_file->value(cfg.tls.cert_file);
    fl_g->input_tls_cert_dir->value(cfg.tls.cert_dir);


    update_samplerates();
    update_channel_lists();

    //fill the DSP section
    fl_g->check_activate_eq->value(cfg.dsp.equalizer);
    
    slider_equalizer1_cb(cfg.dsp.gain1);
    fl_g->equalizerSlider1->value(cfg.dsp.gain1);
    
    slider_equalizer2_cb(cfg.dsp.gain2);
    fl_g->equalizerSlider2->value(cfg.dsp.gain2);
    
    slider_equalizer3_cb(cfg.dsp.gain3);
    fl_g->equalizerSlider3->value(cfg.dsp.gain3);
    
    slider_equalizer4_cb(cfg.dsp.gain4);
    fl_g->equalizerSlider4->value(cfg.dsp.gain4);
    
    slider_equalizer5_cb(cfg.dsp.gain5);
    fl_g->equalizerSlider5->value(cfg.dsp.gain5);
	
	fl_g->check_activate_drc->value(cfg.dsp.compressor);
    fl_g->check_aggressive_mode->value(cfg.dsp.aggressive_mode);
	
	slider_threshold_cb(cfg.dsp.threshold);
    fl_g->thresholdSlider->value(cfg.dsp.threshold);
	
	slider_ratio_cb(cfg.dsp.ratio);
    fl_g->ratioSlider->value(cfg.dsp.ratio);
	
	slider_attack_cb(cfg.dsp.attack);
    fl_g->attackSlider->value(cfg.dsp.attack);
	
	slider_release_cb(cfg.dsp.release);
    fl_g->releaseSlider->value(cfg.dsp.release);
    
	slider_makeup_cb(cfg.dsp.makeup_gain);
    fl_g->makeupSlider->value(cfg.dsp.makeup_gain);
	
    //fill the GUI section
    fl_g->button_gui_bg_color->color(cfg.main.bg_color,
            fl_lighter((Fl_Color)cfg.main.bg_color));
    fl_g->button_gui_text_color->color(cfg.main.txt_color,
            fl_lighter((Fl_Color)cfg.main.txt_color));
    fl_g->check_gui_attach->value(cfg.gui.attach);
    fl_g->check_gui_ontop->value(cfg.gui.ontop);
    if(cfg.gui.ontop)
    {
        fl_g->window_main->stay_on_top(1);
        fl_g->window_cfg->stay_on_top(1);
    }
    fl_g->check_gui_hide_log_window->value(cfg.gui.hide_log_window);
    fl_g->check_gui_remember_pos->value(cfg.gui.remember_pos);

   
    fl_g->check_gui_lcd_auto->value(cfg.gui.lcd_auto);

    fl_g->check_gui_start_minimized->value(cfg.gui.start_minimized);

    
    fl_g->choice_gui_language->value(cfg.gui.lang);
    
    if (cfg.gui.vu_mode == VU_MODE_GRADIENT)
        fl_g->radio_gui_vu_gradient->setonly();
    else
        fl_g->radio_gui_vu_solid->setonly();

}

//Updates the samplerate drop down menu for the audio
//device the user has selected
void update_samplerates(void)
{
    int i;
    int *sr_list;
    char sr_asc[10];

    fl_g->choice_cfg_samplerate->clear();

    sr_list = cfg.audio.pcm_list[cfg.audio.dev_num]->sr_list;

    for(i = 0; sr_list[i] != 0; i++)
    {
        snprintf(sr_asc, sizeof(sr_asc), "%dHz", sr_list[i]);
        fl_g->choice_cfg_samplerate->add(sr_asc);

        if(cfg.audio.samplerate == sr_list[i])
            fl_g->choice_cfg_samplerate->value(i);
    }
    if(i == 0)
    {
        fl_g->choice_cfg_samplerate->add("dev. not supported");
        fl_g->choice_cfg_samplerate->value(0);
    }
}

void update_channel_lists(void)
{
    
    int i;
    char ch_num_txt[16];
    
    int dev_num = fl_g->choice_cfg_dev->value();
    int num_of_channels = cfg.audio.pcm_list[dev_num]->num_of_channels;
    
    fl_g->choice_cfg_left_channel->clear();
    fl_g->choice_cfg_right_channel->clear();
    
    for (i = 1; i <= num_of_channels; i++)
    {
        snprintf(ch_num_txt, sizeof(ch_num_txt), "%d", i);
        fl_g->choice_cfg_left_channel->add(ch_num_txt);
        fl_g->choice_cfg_right_channel->add(ch_num_txt);
    }
    
    fl_g->choice_cfg_left_channel->value(cfg.audio.left_ch-1);
    fl_g->choice_cfg_right_channel->value(cfg.audio.right_ch-1);    
    
}

void print_info(const char* info, int info_type)
{
    char timebuf[10];
    char logtimestamp[21];
    char* infotxt;
    FILE *log_fd;
    int len;

    time_t test;
    struct tm  *mytime;
    static struct tm time_bak;

    infotxt = strdup(info);


    test = time(NULL);
    mytime = localtime(&test);

    if( (time_bak.tm_min != mytime->tm_min) || (time_bak.tm_hour != mytime->tm_hour) )
    {
        time_bak.tm_min = mytime->tm_min;
        time_bak.tm_hour = mytime->tm_hour;
        strftime(timebuf, sizeof(timebuf), "\n%H:%M:", mytime);
        fl_g->info_buffer->append(timebuf);
    }

    fl_g->info_buffer->append((const char*)"\n");
    fl_g->info_buffer->append((const char*)info);

    //always scroll to the last line
    fl_g->info_output->scroll(fl_g->info_buffer->count_lines(0,     //count the lines from char 0 to the last character
                            fl_g->info_buffer->length()),           //returns the number of characters in the buffer
                            0);

    // log to log_file if defined
    if ((cfg.main.log_file != NULL) && (strlen(cfg.main.log_file) > 0))
    {
        strftime(logtimestamp, sizeof(logtimestamp), "%Y-%m-%d %H:%M:%S", mytime);
        log_fd = fl_fopen(cfg.main.log_file, "ab");
        if (log_fd != NULL) 
        {
            if(strchr(infotxt, ':'))
                strrpl(&infotxt, (char*)"\n", (char*)", ", MODE_ALL);
            else
                strrpl(&infotxt, (char*)"\n", (char*)" ", MODE_ALL);

            strrpl(&infotxt, (char*)":,", (char*)": ", MODE_ALL);
            strrpl(&infotxt, (char*)"\t", (char*)"", MODE_ALL);

            len = strlen(infotxt)-1;
            // remove trailing commas and spaces 
            while (infotxt[len] == ',' || infotxt[len] == ' ')
            {
                infotxt[len--] = '\0';
            }

            fprintf(log_fd, "%s %s\n", logtimestamp, infotxt);
            fclose(log_fd);
        }

    }
    
    free(infotxt);
}

void print_lcd(const char *text, int len, int home, int clear)
{
    if(clear)
        fl_g->lcd->clear();

    fl_g->lcd->print((const uchar*)text, len);

    if(home)
        fl_g->lcd->cursor_pos(0);
}


void expand_string(char **str)
{
    int str_len;
    char expanded_str[1024];
    struct tm *date;
    const time_t t = time(NULL);

    // The %i (index number) place holder must be replaced with %%i
    // Otherwise strftime will replace %i with i and the index number will loose its function
    strrpl(str, (char*)"%i", (char*)"%%i", MODE_ALL);
    
    // %c, %x, %X specifiers are not allowed because they return illegal characters for file names
    // Therefore we make sure that strftime will ignore them
    strrpl(str, (char*)"%c", (char*)"%%c", MODE_ALL);
    strrpl(str, (char*)"%x", (char*)"%%x", MODE_ALL);
    strrpl(str, (char*)"%X", (char*)"%%X", MODE_ALL);

    date = localtime(&t);
    strftime(expanded_str, sizeof(expanded_str)-1, *str, date);
    
    str_len = strlen(expanded_str);
    *str = (char *)realloc(*str, str_len+1);
    strncpy(*str, expanded_str, str_len+1);
}

void test_file_extension(void)
{
    char *current_ext;

    current_ext = util_get_file_extension(cfg.rec.filename);

    // Append extension
    if(current_ext == NULL)
    {
        cfg.rec.filename = (char*)realloc(cfg.rec.filename, strlen(cfg.rec.filename)+strlen(cfg.rec.codec)+2);
        strcat(cfg.rec.filename, ".");
        strcat(cfg.rec.filename, cfg.rec.codec);
        fl_g->input_rec_filename->value(cfg.rec.filename);
    }
    // Replace extension
    else if(strcmp(current_ext, cfg.rec.codec))
    {
        strrpl(&cfg.rec.filename, current_ext, cfg.rec.codec, MODE_LAST);
        fl_g->input_rec_filename->value(cfg.rec.filename);
    }
}



void init_main_gui_and_audio(void)
{
    if(cfg.gui.remember_pos)
    {
        int x, y, w, h;
        x = cfg.gui.x_pos;
        y = cfg.gui.y_pos;
        w = fl_g->window_main->w();
        h = fl_g->window_main->h();
        
        // Move butt window to the saved position only if at least 50 pixel of butt are visible on the screen
        if( (x+w > 50) && (x+50 < Fl::w()) && (y+h > 50) && (y+50 < Fl::h()) )
            fl_g->window_main->position(cfg.gui.x_pos, cfg.gui.y_pos);
    }
    
    fl_g->slider_gain->value(util_factor_to_db(cfg.main.gain));
    fl_g->window_main->redraw();
    
    if(cfg.gui.ontop)
        fl_g->window_main->stay_on_top(1);
    
    fl_g->button_info->label(_("Hide log"));
    if(cfg.gui.hide_log_window)
       button_info_cb();
          

    lame_stream.channel = cfg.audio.channel;
    lame_stream.bitrate = cfg.audio.bitrate;
    lame_stream.samplerate = cfg.audio.samplerate;
    lame_enc_reinit(&lame_stream);

    lame_rec.channel = cfg.audio.channel;
    lame_rec.bitrate = cfg.rec.bitrate;
    lame_rec.samplerate = cfg.audio.samplerate;
    lame_enc_reinit(&lame_rec);

    vorbis_stream.channel = cfg.audio.channel;
    vorbis_stream.bitrate = cfg.audio.bitrate;
    vorbis_stream.samplerate = cfg.audio.samplerate;
    vorbis_enc_reinit(&vorbis_stream);

    vorbis_rec.channel = cfg.audio.channel;
    vorbis_rec.bitrate = cfg.rec.bitrate;
    vorbis_rec.samplerate = cfg.audio.samplerate;
    vorbis_enc_reinit(&vorbis_rec);
    
    opus_stream.channel = cfg.audio.channel;
    opus_stream.bitrate = cfg.audio.bitrate*1000;
    opus_stream.samplerate = cfg.audio.samplerate;
    opus_enc_reinit(&opus_stream);
    
    opus_rec.channel = cfg.audio.channel;
    opus_rec.bitrate = cfg.rec.bitrate*1000;
    opus_rec.samplerate = cfg.audio.samplerate;
    opus_enc_reinit(&opus_rec);

#ifdef HAVE_LIBFDK_AAC
    if (g_aac_lib_available == 1)
    {
        aac_stream.channel = cfg.audio.channel;
        aac_stream.bitrate = cfg.audio.bitrate;
        aac_stream.samplerate = cfg.audio.samplerate;
        aac_stream.aot = cfg.audio.aac_aot;
        aac_enc_reinit(&aac_stream);
        
        aac_rec.channel = cfg.audio.channel;
        aac_rec.bitrate = cfg.rec.bitrate;
        aac_rec.samplerate = cfg.audio.samplerate;
        aac_rec.aot = cfg.audio.aac_aot;
        aac_enc_reinit(&aac_rec);
    }
#endif
    
    flac_stream.channel = cfg.audio.channel;
    flac_stream.samplerate = cfg.audio.samplerate;
    flac_stream.enc_type = FLAC_ENC_TYPE_STREAM;
    flac_enc_reinit(&flac_stream);

    flac_rec.channel = cfg.audio.channel;
    flac_rec.samplerate = cfg.audio.samplerate;
    flac_rec.enc_type = FLAC_ENC_TYPE_REC;
    flac_enc_reinit(&flac_rec);
}


char *ask_user_msg = NULL;
char *ask_user_hash = NULL;
int ask_user_has_clicked = 0;
int ask_user_answer = 0;

void ask_user_reset(void)
{
    if (ask_user_msg != NULL)
    {
        free(ask_user_msg);
        ask_user_msg = NULL;
    }
    if (ask_user_hash != NULL)
    {
        free(ask_user_hash);
        ask_user_hash = NULL;
    }
    ask_user_has_clicked = 0;
}

int ask_user_get_has_clicked(void)
{
    return ask_user_has_clicked;
}

int ask_user_get_answer(void)
{
    return ask_user_answer;
}

void ask_user_set_msg(char *msg)
{
    int len = strlen(msg);
    ask_user_msg = (char*)calloc(len+1, sizeof(char));
    strncpy(ask_user_msg, msg, len);
}

void ask_user_set_hash(char *hash)
{
    int len = strlen(hash);
    ask_user_hash = (char*)calloc(len+1, sizeof(char));
    strncpy(ask_user_hash, hash, len);
}

void ask_user_ask(void)
{
    if (fl_choice("%s", _("TRUST"), _("No"), NULL, ask_user_msg) == 1)
    { // No
        ask_user_answer = IC_ABORT;
    }
    else
    { // TRUST
	int len = strlen(ask_user_hash);
        cfg.srv[cfg.selected_srv]->cert_hash =
            (char*)realloc(cfg.srv[cfg.selected_srv]->cert_hash, len+1);
        
        memset(cfg.srv[cfg.selected_srv]->cert_hash, 0, len+1);
        strncpy(cfg.srv[cfg.selected_srv]->cert_hash, ask_user_hash, len);
        ask_user_answer = IC_RETRY;
    }
    
    ask_user_has_clicked = 1;
}
